File: week01/example/ltc155_min_stack.cpp
//
// Created by Opt_c on 2022/4/18.
//
#include <bits/stdc++.h>

using namespace std;

class MinStack {
public:
    MinStack() {

    }

    void push(int val) {
        a.push(val);
        if (preMin.empty()) preMin.push(val);
        else preMin.push(min(preMin.top(), val));
    }

    void pop() {
        a.pop();
        preMin.pop();

    }

    int top() {
        return a.top();
    }

    int getMin() {
        return preMin.top();
    }

private:
    stack<int> a;
    stack<int> preMin;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: week01/example/ltc155_min_stack.py
class MinStack:

    def __init__(self):
        self.stack = list()
        self.preMin_stack = list()

    def push(self, val: int) -> None:
        self.stack.append(val)
        if self.preMin_stack:
            self.preMin_stack.append(min(self.preMin_stack[-1], val))
        else:
            self.preMin_stack.append(val)

    def pop(self) -> None:
        self.stack.pop()
        self.preMin_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.preMin_stack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

if __name__ == '__main__':
    obj = MinStack()
    obj.push(3)
    obj.push(4)
    obj.push(5)
    obj.push(6)
    obj.pop()
    param_3 = obj.top()
    param_4 = obj.getMin()
